src/abstractdataframe/reshape.jl

```diff
@@ -121,137 +198,240 @@ function _stackview(df::AbstractDataFrame, measure_vars::AbstractVector{Int},
+├─────┼───────┼───────┼───────┼───────────┼───────────┤
+│ 1   │ 1     │ 1     │ 1     │ 1.20649   │ -1.27628  │
+│ 2   │ 2     │ 1     │ 1     │ -0.917794 │ 0.940007  │
+│ 3   │ 3     │ 2     │ 1     │ 0.309629  │ 0.820397  │
+│ 4   │ 4     │ 2     │ 2     │ 1.46677   │ -1.03457  │
+│ 5   │ 5     │ 3     │ 2     │ 1.04339   │ -0.770464 │
+│ 6   │ 6     │ 3     │ 2     │ -0.172475 │ -2.81039  │
+
+julia> unstack(long, :id, :variable, :value)
+6×3 DataFrame
+│ Row │ id    │ c         │ d         │
+│     │ Int64 │ Float64?  │ Float64?  │
+├─────┼───────┼───────────┼───────────┤
+│ 1   │ 1     │ 1.20649   │ -1.27628  │
+│ 2   │ 2     │ -0.917794 │ 0.940007  │
+│ 3   │ 3     │ 0.309629  │ 0.820397  │
+│ 4   │ 4     │ 1.46677   │ -1.03457  │
+│ 5   │ 5     │ 1.04339   │ -0.770464 │
+│ 6   │ 6     │ -0.172475 │ -2.81039  │
+
+julia> unstack(long, [:id, :a], :variable, :value)
+6×4 DataFrame
+│ Row │ id    │ a     │ c         │ d         │
+│     │ Int64 │ Int64 │ Float64?  │ Float64?  │
+├─────┼───────┼───────┼───────────┼───────────┤
+│ 1   │ 1     │ 1     │ 1.20649   │ -1.27628  │
+│ 2   │ 2     │ 1     │ -0.917794 │ 0.940007  │
+│ 3   │ 3     │ 2     │ 0.309629  │ 0.820397  │
+│ 4   │ 4     │ 2     │ 1.46677   │ -1.03457  │
+│ 5   │ 5     │ 3     │ 1.04339   │ -0.770464 │
+│ 6   │ 6     │ 3     │ -0.172475 │ -2.81039  │
+
+julia> unstack(long, :id, :variable, :value, renamecols=x->Symbol(:_, x))
+6×3 DataFrame
+│ Row │ id    │ _c        │ _d        │
+│     │ Int64 │ Float64?  │ Float64?  │
+├─────┼───────┼───────────┼───────────┤
+│ 1   │ 1     │ 1.20649   │ -1.27628  │
+│ 2   │ 2     │ -0.917794 │ 0.940007  │
+│ 3   │ 3     │ 0.309629  │ 0.820397  │
+│ 4   │ 4     │ 1.46677   │ -1.03457  │
+│ 5   │ 5     │ 1.04339   │ -0.770464 │
+│ 6   │ 6     │ -0.172475 │ -2.81039  │
 ```
 Note that there are some differences between the widened results above.
 """
 function unstack(df::AbstractDataFrame, rowkey::ColumnIndex, colkey::ColumnIndex,
-                 value::ColumnIndex; renamecols::Function=identity)
-    refkeycol = categorical(df[!, rowkey])
-    droplevels!(refkeycol)
-    keycol = categorical(df[!, colkey])
-    droplevels!(keycol)
+                 value::ColumnIndex; renamecols::Function=identity,
+                 allowmissing::Bool=false, allowduplicates::Bool=false)
+    refkeycol = df[!, rowkey]
+    keycol = df[!, colkey]
     valuecol = df[!, value]
     return _unstack(df, index(df)[rowkey], index(df)[colkey],
-                    keycol, valuecol, refkeycol, renamecols)
+                    keycol, valuecol, refkeycol, renamecols, allowmissing, allowduplicates)
+end
+
+function _unstack_preprocess_vector(v::AbstractVector)
+    v_unique = unique(v)
+    had_missing = any(ismissing, v_unique)
+    v_unique = intersect(levels(v), v_unique)
+    had_missing && (v_unique = vcat(v_unique, [missing]))
+    len_v = length(v_unique)
+    v_map = Dict([x => i for (i,x) in enumerate(v_unique)])
```

Comment:
How hard do you think it would be to use `refpool` here when it's defined, and work with `refarray` for the rest of the processing? That should be much faster for these types (and avoid a regression for `CategoricalArray` and `PooledArrays`).

Reply:
Earlier we did:
```
    refkeycol = categorical(df[!, rowkey])
    droplevels!(refkeycol)
    keycol = categorical(df[!, colkey])
    droplevels!(keycol)
```
which I think is also expensive (please correct me if I am wrong - you essentially have to do a very similar lookup to the one that I propose now).

The point is that we have to make sure that the mapping is to values `1,2,...` in the right order (and that is why `droplevels!` is needed and in general `PooledArray` is not useful as it can contain "holes" in level numbers and no order of them is guaranteed).

The only case when it could be potentially faster if when we already have a `CategoricalVector` that has dropped levels (but even then I am not sure if `droplevels!` is cost-free - you know the internals better here).

Reply:
The point is that when the input was a `CategoricalArray` or `PooledArray`, the `categorical` call operated directly on the reference codes, which are much faster to deal with than performing a dict lookup for each value (we look for the new reference code in a vector indexed by old reference codes). And in the ideal case where all levels were used and in the right order (which is quite common at least for `CategoricalArray`), the call only implied doing a (very fast) copy.

Reply:
Here is a benchmark showing the performance penalty of what I proposed:
```
wide = DataFrame(id = 1:6000000,
                        a  = repeat([1:3;], inner = [2000000]),
                        b  = repeat([1:2;], inner = [3000000]),
                        c  = randn(6000000),
                        d  = randn(6000000));

long = stack(wide);
@time unstack(long, :id, :variable, :value); # master: 9.839218 seconds (6.00 M allocations: 2.726 GiB, 8.00% gc time)
                                             # this PR: 15.094627 seconds (132.00 M allocations: 4.858 GiB, 4.90% gc time)
categorical!(long, 1:4);
@time unstack(long, :id, :variable, :value); # master: 9.277382 seconds (18.00 M allocations: 3.044 GiB, 32.15% gc time)
                                             # this PR: 15.921799 seconds (126.99 M allocations: 4.883 GiB, 10.74% gc time)
```

Reply:
Yeah so about 80% slower.

Do you think it would be hard to use `refpool`? If `_unstack_preprocess_vector` returned `keycol_map = refpool(keycol)`, and `rvkeycol = refvalue(keycol)` to be used as `rvkeycol[k]` instead of `keycol[k]`, wouldn't everything work?

Reply:
> If `_unstack_preprocess_vector` returned `keycol_map = refpool(keycol)`

It would work if `refpool(keycol)` guaranteed that the values that are present are mapped to numbers 1,2,... in a continuous way. Unfortunately neither `PolledArray` nor `CategoricalArray` guarantee this. But freshly allocated `PooledArray` ensures such behavior and this is what I will try to implement.

Reply:
@nalimilan - I have pushed a `PooledVector` implementation. It is much faster than what we have on master (the first non-categorical test is ~2 seconds). Can you please have a look if you think it is OK? I have still one problem with CategoricalArrays.jl though. I will comment in the code.